### PR TITLE
Adding more stuff to GWAS Catalog ingest

### PR DESCRIPTION
**Pull request by DSuveges**
- [x] Draft version of the documentation added.
- [x] Slight update on the logic added (Adding AF from variant annotation).
- [x] Deduplicating association 2 variant mapping based on allele frequency. (Keeping always the variant with the highest MAF, given gwas-es are focusing on common variants)
- [x] Harmonizing effect (calculate OR/beta, Z-score, direction of effect etc.)
- [ ] Ingest GWAS Catalog Study table.
- [ ] Merge with associations
- [ ] Split studies where required.

### CODE REVIEW

**Comment by DSuveges** [comment]
I would get this PR reviewed and merged before it grows too large to realistically review. However in the meantime I'm working on the completion of the GWAS Catalog ingestion and there will be an other PR soon.

File: src/etl/gwas_ingest/process_associations.py [hunk 1]
```diff
@@ -37,6 +37,47 @@
...
+    with the highest minor allele frequency from all mappings. This process is population name agnostic,
+    the only assumption is the field name is `allele_frequencies` and it is a structType.
+
+    Args:
+      df (DataFrame): DataFrame
+
+    Returns:
+      A DataFrame with identical columns
+    """
+    # parsing population names from schema:
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/process_associations.py:52]: Not clear to me how this info is encoded. Are the populations inside an array of structs?

File: src/etl/gwas_ingest/process_associations.py [hunk 2]
```diff
@@ -37,6 +37,47 @@
...
+    pop_names = [
+        field for field in df.schema.fields if field.name == "allele_frequencies"
+    ][0].dataType.fieldNames()
+
+    def af2maf(c: Column) -> Column:
+        """Column function to calculate minor allele frequency from allele frequency"""
+        return f.when(c > 0.5, 1 - c).otherwise(c)
+
+    # Windowing through all associations. Within an associations, rows are ordered by the maximum MAF:
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/process_associations.py:61]:
```suggestion
    # Windowing through all associations. Within an association, rows are ordered by the maximum MAF:
```

File: src/etl/gwas_ingest/process_associations.py [hunk 3]
```diff
@@ -37,6 +37,47 @@
...
+        .withColumn("row_number", f.row_number().over(w))
+        .filter(f.col("row_number") == 1)
+        .drop("row_number")
+        .persist()
+    )
+
+
 def concordance_filter(df: DataFrame) -> DataFrame:
     """
     This function filters for variants with concordant alleles witht the association's reported risk allele.
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/process_associations.py:83]:
```suggestion
    This function filters for variants with concordant alleles with the association's reported risk allele.
```

File: tests/test_effect_harmonization.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+if TYPE_CHECKING:
+    from pyspark.sql import DataFrame, SparkSession
+
+
+@pytest.fixture
+def mock_allele_columns(spark: SparkSession) -> DataFrame:
+    return spark.createDataFrame(
+        [
+            {"allele": "A", "reverseComp": "T", "isPalindrom": False},
```
Comment (ireneisdoomed) [inline tests/test_effect_harmonization.py:18]: There's a typo in the name of the column
```suggestion
            {"allele": "A", "reverseComp": "T", "isPalindrome": False},
```

**Review by ireneisdoomed — APPROVED** [review]
This is great. I think all the logic makes sense, just added minor comments.

File: src/etl/gwas_ingest/effect_harmonization.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            f"parsed_${pvalcol}",
+            f.when(f.col(f"parsed_${pvalcol}") == 0, sys.float_info.min).otherwise(
+                f.col(f"parsed_${pvalcol}")
+            ),
+        )
+        # Convert pvalues to z-score:
+        .withColumn(
+            "zscore",
+            f.udf(
+                lambda pv: float(abs(st.norm.ppf(pv / 2))) if pv else None,
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/effect_harmonization.py:40]: As this package is in the base Python library from Python 3.8, we should use it even if it was not as performant.

File: src/etl/gwas_ingest/effect_harmonization.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                    f.col("confidence_interval").contains("decrease")
+                    & ~f.col("needs_harmonization")
+                ),
+                f.col("beta") * -1,
+            ).otherwise(f.col("beta")),
+        )
+        .withColumn(
+            "beta_conf_intervals",
+            f.array(
+                f.col("beta") - f.lit(1.96) * f.col("beta") / f.col("zscore"),
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/effect_harmonization.py:109]: OK, I'm adding it with the next commit.

File: src/etl/gwas_ingest/process_associations.py (hunk 1 above)
Comment (DSuveges) [reply src/etl/gwas_ingest/process_associations.py:52]: Yeah, I shouldn't have relied on the automatic document generation. I'll clarify.

File: tests/test_effect_harmonization.py (hunk 4 above)
Comment (DSuveges) [reply tests/test_effect_harmonization.py:18]: I don't know how it could be passed. Thank you for noticing.

**Pull request merged**